src/robin_dict.jl

```diff
@@ -0,0 +1,611 @@
+    end
+    return index_curr
+end
+
+function rh_insert_for_rehash!(h_new::RobinDict{K, V}, key::K, val::V, hash::UInt32) where {K, V}
+    # table full
+    @assert h_new.count != length(h_new.keys)
+    
+    ckey, cval, chash = key, val, hash
+    sz = length(h_new.keys)
+    index_init = desired_index(chash, sz)
+
+    index_curr = index_init
+    probe_distance = 0
+    probe_current = 0
+    @inbounds while true
+        if (isslotempty(h_new, index_curr))
+            break
+        end
+        probe_distance = calculate_distance(h_new, index_curr)
+
+        if probe_current > probe_distance
+            h_new.vals[index_curr], cval = cval, h_new.vals[index_curr]
+            h_new.keys[index_curr], ckey = ckey, h_new.keys[index_curr]
+            h_new.hashes[index_curr], chash = chash, h_new.hashes[index_curr]
+            probe_current = probe_distance
+        end
+        probe_current += 1
+        index_curr = (index_curr & (sz - 1)) + 1
+    end
+
+    @inbounds if isslotempty(h_new, index_curr)
+        h_new.count += 1
+    end
+
+    @inbounds h_new.vals[index_curr] = cval
+    @inbounds h_new.keys[index_curr] = ckey
+    @inbounds h_new.hashes[index_curr] = chash
+    
+    @assert probe_current >= 0
+    
+    if h_new.idxfloor == 0
+        h_new.idxfloor = index_curr
+    else
+        h_new.idxfloor = min(h_new.idxfloor, index_curr)
+    end
+    return index_curr
+end
+
+#rehash! algorithm
+function rehash!(h::RobinDict{K,V}, newsz = length(h.keys)) where {K, V}
+    oldk = h.keys
+    oldv = h.vals
+    oldh = h.hashes
+    sz = length(oldk)
+    newsz = _tablesz(newsz)
+    if h.count == 0
+        resize!(h.keys, sz)
+        resize!(h.vals, sz)
+        resize!(h.hashes, newsz)
+        fill!(h.hashes, 0)
+        h.count = 0
+        h.idxfloor = 0
+        return h
+    end
+
+    h.keys = Vector{K}(undef, newsz)
+    h.vals = Vector{V}(undef, newsz)
+    h.hashes = zeros(UInt32,newsz)
+    h.count = 0
+    h.idxfloor = 0
+
+    for i = 1:sz
+        @inbounds if oldh[i] != 0
+            k = oldk[i]
+            v = oldv[i]
+            rh_insert_for_rehash!(h, k, v, oldh[i])
+        end
+    end
+    return h
+end
+
+function sizehint!(d::RobinDict, newsz)
+    newsz = _tablesz(newsz*2)  # *2 for keys and values in same array
+    oldsz = length(d.keys)
+    # grow at least 25%
+    if newsz < (oldsz*5)>>2
+        return d
+    end
+    rehash!(d, newsz)
+end
+
+@propagate_inbounds isslotfilled(h::RobinDict, index) = (h.hashes[index] != 0)
+@propagate_inbounds isslotempty(h::RobinDict, index) = (h.hashes[index] == 0)
+
+
+function setindex!(h::RobinDict{K,V}, v0, key0) where {K, V}
+    key = convert(K, key0)
+    isequal(key, key0) || throw(ArgumentError("$key0 is not a valid key for type $K"))
+    _setindex!(h, key, v0)
+end
+
+function _setindex!(h::RobinDict{K,V}, key::K, v0) where {K, V}
+    v = convert(V, v0)
+    sz = length(h.keys)
+    (h.count > ROBIN_DICT_LOAD_FACTOR * sz) && rehash!(h, sz<<2)
```

Comment:
Always resizing by a factor of 4 looks sketchy (leads to very low load factors and hence memory waste for long-lived dicts). RobinDict should now be much cheaper to rehash than Base dict.

Maybe just copy what Base does? (i.e. use threshold to decide whether to grow by factor 2 or 4)

Reply:
I did that on purpose as this change showed some improvement in benchmarks . In details described [here](https://nextjournal.com/eulerkochy/blog-two-jsoc/)